txnkv/transaction/pessimistic.go

```diff
@@ -150,156 +168,378 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
+		if len(locks) > 0 {
+			// Because we already waited on tikv, no need to Backoff here.
+			// tikv default will wait 3s(also the maximum wait value) when lock error occurs
+			if diagCtx.resolvingRecordToken == nil {
+				token := c.store.GetLockResolver().RecordResolvingLocks(locks, c.startTS)
+				diagCtx.resolvingRecordToken = &token
 			} else {
-				// the lockWaitTime is set, we should return wait timeout if we are still blocked by a lock
-				if time.Since(lockWaitStartTime).Milliseconds() >= action.LockWaitTime() {
-					return errors.WithStack(tikverr.ErrLockWaitTimeout)
-				}
+				c.store.GetLockResolver().UpdateResolvingLocks(locks, c.startTS, *diagCtx.resolvingRecordToken)
 			}
-			if action.LockCtx.PessimisticLockWaited != nil {
-				atomic.StoreInt32(action.LockCtx.PessimisticLockWaited, 1)
+			resolveLockOpts := txnlock.ResolveLocksOptions{
+				CallerStartTS: 0,
+				Locks:         locks,
+			}
+			if action.LockCtx.Stats != nil {
+				resolveLockOpts.Detail = &action.LockCtx.Stats.ResolveLock
+			}
+			resolveLockRes, err := c.store.GetLockResolver().ResolveLocksWithOpts(bo, resolveLockOpts)
+			if err != nil {
+				return true, nil, err
 			}
-		}
 
-		// Handle the killed flag when waiting for the pessimistic lock.
-		// When a txn runs into LockKeys() and backoff here, it has no chance to call
-		// executor.Next() and check the killed flag.
-		if action.Killed != nil {
-			// Do not reset the killed flag here!
-			// actionPessimisticLock runs on each region parallelly, we have to consider that
-			// the error may be dropped.
-			if atomic.LoadUint32(action.Killed) == 1 {
-				return errors.WithStack(tikverr.ErrQueryInterrupted)
+			// If msBeforeTxnExpired is not zero, it means there are still locks blocking us acquiring
+			// the pessimistic lock. We should return acquire fail with nowait set or timeout error if necessary.
+			if resolveLockRes.TTL > 0 {
+				if action.LockWaitTime() == kv.LockNoWait {
+					return true, nil, errors.WithStack(tikverr.ErrLockAcquireFailAndNoWaitSet)
+				} else if action.LockWaitTime() == kv.LockAlwaysWait {
+					// do nothing but keep wait
+				} else {
+					// the lockWaitTime is set, we should return wait timeout if we are still blocked by a lock
+					if time.Since(action.WaitStartTime).Milliseconds() >= action.LockWaitTime() {
+						return true, nil, errors.WithStack(tikverr.ErrLockWaitTimeout)
+					}
+				}
+				if action.LockCtx.PessimisticLockWaited != nil {
+					atomic.StoreInt32(action.LockCtx.PessimisticLockWaited, 1)
+				}
 			}
+			return false, &failedMutations, nil
 		}
+
+		return true, nil, errors.New("Pessimistic lock response corrupted")
+	}
+
+	if regionErr != nil || len(locks) != 0 {
+		return true, nil, errors.New("Pessimistic lock response corrupted")
+	}
+
+	// If the Results in response, there must be either some unretryable error in keyErrs or some region error, therefore
+	// the function must have returned.
+	if len(lockResp.Results) == 0 {
```

Comment:
This comment is a little bit confusing, why the `len(lockResp.Results)` must be zero here?

Reply:
Sorry the comment is mistakenly imcomplete. This should be an unreachable path. I'll update it.